src/ipvs/ip_vs_mh.c

```diff
@@ -0,0 +1,527 @@
+ *
+ * The mh algorithm is to assign a preference list of all the lookup
+ * table positions to each destination and populate the table with
+ * the most-preferred position of destinations. Then it is to select
+ * destination with the hash key of source IP address through looking
+ * up a the lookup table.
+ *
+ * The algorithm is detailed in:
+ * [3.4 Consistent Hasing]
+ * https://www.usenix.org/system/files/conference/nsdi16/nsdi16-paper-eisenbud.pdf
+ *
+ * see net/netfilter/ipvs/ip_vs_mh.c for reference
+ */
+#include "ipvs/mh.h"
+#include "ipvs/siphash.h"
+#include "ipvs/kcompat.h"
+#include "conf/inet.h"
+#include "conf/dest.h"
+
+#define DP_VS_SVC_F_SCHED_MH_FALLBACK   DP_VS_SVC_F_SCHED1 /* MH fallback */
+#define DP_VS_SVC_F_SCHED_MH_PORT       DP_VS_SVC_F_SCHED2 /* MH use port */
+
+struct dp_vs_mh_lookup {
+    struct dp_vs_dest *dest; /* real server (cache) */
+};
+
+struct dp_vs_mh_dest_setup {
+    unsigned int offset; /* starting offset */
+    unsigned int skip; /* skip */
+    unsigned int perm; /* next_offset */
+    int turns; /* weight / gcd() and rshift */
+};
+
+/* Available prime numbers for MH table */
+static int primes[] = {251, 509, 1021, 2039, 4093,
+    8191, 16381, 32749, 65521, 131071};
+
+/* For DPVS MH entry hash table */
+#ifndef CONFIG_DP_VS_MH_TAB_INDEX
+#define CONFIG_DP_VS_MH_TAB_INDEX   12
+#endif
+#define DP_VS_MH_TAB_BITS           (CONFIG_DP_VS_MH_TAB_INDEX / 2)
+#define DP_VS_MH_TAB_INDEX          (CONFIG_DP_VS_MH_TAB_INDEX - 8)
+#define DP_VS_MH_TAB_SIZE           primes[DP_VS_MH_TAB_INDEX]
+
+struct dp_vs_mh_state {
+    struct dp_vs_mh_lookup *lookup;
+    struct dp_vs_mh_dest_setup *dest_setup;
+    hsiphash_key_t hash1, hash2;
+    int gcd;
+    int rshift;
+};
+
+int dp_vs_fill_iphdr(int af, struct rte_mbuf *mbuf,
+                                   struct dp_vs_iphdr *iph);
+
+static inline void generate_hash_secret(hsiphash_key_t *hash1,
+        hsiphash_key_t *hash2)
+{
+    hash1->key[0] = 2654435761UL;
+    hash1->key[1] = 2654435761UL;
+
+    hash2->key[0] = 2654446892UL;
+    hash2->key[1] = 2654446892UL;
+}
+
+/* Helper function to determine if server is unavailable */
+static inline bool is_unavailable(struct dp_vs_dest *dest)
+{
+    return rte_atomic16_read(&dest->weight) <= 0 ||
+        dest->flags & DPVS_DEST_F_OVERLOAD;
+}
+
+/* Returns hash value for IPVS MH entry */
+static inline unsigned int dp_vs_mh_hashkey(int af, const union inet_addr *addr,
+        __be16 port, hsiphash_key_t *key, unsigned int offset)
+{
+    unsigned int v;
+    __be32 addr_fold = inet_addr_fold(af, addr);
+
+    v = (offset + ntohs(port) + ntohl(addr_fold));
+    return hsiphash(&v, sizeof(v), key);
+}
+
+/* Reset all the hash buckets of the specified table. */
+static void dp_vs_mh_reset(struct dp_vs_mh_state *s)
+{
+    int i;
+    struct dp_vs_mh_lookup *l;
+
+    l = &s->lookup[0];
+    for (i = 0; i < DP_VS_MH_TAB_SIZE; i++) {
+        if (l->dest)
+            l->dest = NULL;
+        l++;
+    }
```

Comment:
The object pointed to by l-> dest will not be freed in advance; because deleting rs will first remove rs from the rs linked list of svc, and then regenerate the lookup array based on the rs linked list of svc, and then release the rs space.

Reply:
I will add  a reference count for the set and unset of l->dest